webots_ros2_core/webots_ros2_core/devices/device_manager.py

```diff
@@ -48,29 +50,100 @@ def __init__(self, node, config=None):
-                    node, wb_device, config.get(wb_device.getName(), None))
+                device = DistanceSensorDevice(node, wb_device, self.__config.get(wb_device.getName(), None))
             elif wb_device.getNodeType() == Node.LIGHT_SENSOR:
-                self._devices[wb_device.getName()] = LightSensorDevice(node, wb_device, config.get(wb_device.getName(), None))
+                device = LightSensorDevice(node, wb_device, self.__config.get(wb_device.getName(), None))
+
+            # Add device to the list
+            self.__wb_devices[wb_device.getName()] = wb_device
+            if device:
+                self.__devices[wb_device.getName()] = device
+
+        # Multi-Webots-device (insert if not configured + create configured)
+        self.__insert_imu_device()
+        for config_key in self.__config.keys():
+            if self.__is_imu_device(config_key):
+                self.__devices[config_key] = ImuDevice(node, self.__get_imu_wb_devices_from_key(config_key))
 
         # Verify parameters
-        for device_name in config.keys():
-            if device_name not in self._devices and device_name not in ['@auto']:
-                self._node.get_logger().warn(f'There is no device with name `{device_name}`')
+        for device_name in self.__config.keys():
+            if device_name not in ['@auto'] and device_name not in self.__devices.keys():
+                self.__node.get_logger().warn(
+                    f'Device `{device_name}` is has not considered! The device doesn\'t exist or it is not supported.')
 
         # Create a loop
-        self._node.create_timer(1e-3 * int(node.robot.getBasicTimeStep()), self._callback)
+        self.__node.create_timer(1e-3 * int(node.robot.getBasicTimeStep()), self.__callback)
 
-    def _callback(self):
-        for device in self._devices.values():
+    def __callback(self):
+        for device in self.__devices.values():
             device.step()
+
+    def __is_imu_device(self, device_key):
+        return any(self.__get_imu_wb_devices_from_key(device_key))
+
+    def __get_imu_wb_devices_from_key(self, device_key):
+        wb_device_names = device_key.split('|')
+
+        accelerometer = None
+        inertial_unit = None
+        gyro = None
+
+        for wb_device_name in wb_device_names:
+            if wb_device_name in self.__wb_devices:
+                if self.__wb_devices[wb_device_name] == Node.ACCELEROMETER:
+                    accelerometer = self.__wb_devices[wb_device_name]
+                elif self.__wb_devices[wb_device_name] == Node.INERTIAL_UNIT:
+                    inertial_unit = self.__wb_devices[wb_device_name]
+                elif self.__wb_devices[wb_device_name] == Node.GYRO:
+                    gyro = self.__wb_devices[wb_device_name]
+
+        return [accelerometer, gyro, inertial_unit]
+
+    def __insert_imu_device(self):
+        """Insert Imu device only if non is configured and there is only one in the robot."""
```

Comment:
if non is configured ? do you mean none?

Reply:
I've just improved it, maybe it is more clear now. Basically, there are two conditions that have to be fulfilled before automatic device merging (of Accelerometer, Gyro and InertialUnit) is initiated:
- User didn't define any configuration regarding Imu and
- There is no more than one instance of Accelerometer, Gyro or InertialUnit.

Those conditions are fulfilled for the majority of the robots, but if they are not fulfilled it means that the user expects a custom behaviour.

It was a bit tricky to design this, but I am happy to change it if there is a better idea :)

Reply:
> It was a bit tricky to design this, but I am happy to change it if there is a better idea

Yes, I believe you, this is probably the most complex sensor.
About the multiple Accelerometer, Gyro and InertialUnit case, we could use the position extracted from the URDF to group them correctly together, but the current solution is already very good and works perfectly for most of the cases as you said, so let's keep it like this and extend it in the future if we need.